File: kirun-js/__tests__/engine/function/system/date/SubtractTimeTest.ts
import { FunctionExecutionParameters, KIRunFunctionRepository, KIRunSchemaRepository, Namespaces } from "../../../../../src";
import { AbstractDateFunction } from "../../../../../src/engine/function/system/date/AbstractDateFunction";
import { DateFunctionRepository } from "../../../../../src/engine/function/system/date/DateFunctionRepository";


const dfr : DateFunctionRepository = new DateFunctionRepository();

const fep : FunctionExecutionParameters = new FunctionExecutionParameters(
    new KIRunFunctionRepository(),
    new KIRunSchemaRepository());

test('check for invalid dates', async () => {

    const subtractTimeFunction = await dfr.find(Namespaces.DATE, 'SubtractTime');

    if (!subtractTimeFunction) {
        throw new Error("Function not found");
    }
   
    fep.setArguments(new Map<string, any>([
        [AbstractDateFunction.PARAMETER_DATE_NAME, '2029-15-05T06:04:18.073Z'],
        [AbstractDateFunction.PARAMETER_INT_NAME, 10],
        [AbstractDateFunction.PARAMETER_UNIT_NAME, 'SECOND']
    ]));

    await expect( () => subtractTimeFunction.execute(fep)).rejects.toThrow();
});

test('Subtract Time 1', async () => {

    const subtractTimeFunction = await dfr.find(Namespaces.DATE, 'SubtractTime');

    if (!subtractTimeFunction) {
        throw new Error("Function not found");
    }
    
    fep.setArguments(new Map<string, any>([
        [AbstractDateFunction.PARAMETER_DATE_NAME, '2024-09-13T23:52:34.633-05:30'],
        [AbstractDateFunction.PARAMETER_INT_NAME, 10],
        [AbstractDateFunction.PARAMETER_UNIT_NAME, 'MINUTE']
    ]));

    expect((await subtractTimeFunction.execute(fep)).allResults()[0].getResult().get('result')).toBe('2024-09-13T23:42:34.633-05:30');
});

test('Subtract Time 2', async () => {

    const subtractTimeFunction = await dfr.find(Namespaces.DATE, 'SubtractTime');

    if (!subtractTimeFunction) {
        throw new Error("Function not found");
    }
    
    fep.setArguments(new Map<string, any>([
        [AbstractDateFunction.PARAMETER_DATE_NAME, '2024-09-13T23:52:34.633-05:30'],
        [AbstractDateFunction.PARAMETER_INT_NAME, 13],
        [AbstractDateFunction.PARAMETER_UNIT_NAME, 'MONTH']
    ]));

    expect((await subtractTimeFunction.execute(fep)).allResults()[0].getResult().get('result')).toBe('2023-08-13T23:52:34.633-05:30');
});

test('Subtract Time 3', async () => {

    const subtractTimeFunction = await dfr.find(Namespaces.DATE, 'SubtractTime');

    if (!subtractTimeFunction) {
        throw new Error("Function not found");
    }
    
    fep.setArguments(new Map<string, any>([
        [AbstractDateFunction.PARAMETER_DATE_NAME, '2024-09-13T23:52:34.633-05:30'],
        [AbstractDateFunction.PARAMETER_INT_NAME, 70],
        [AbstractDateFunction.PARAMETER_UNIT_NAME, 'SECOND']
    ]));

    expect((await subtractTimeFunction.execute(fep)).allResults()[0].getResult().get('result')).toBe('2024-09-13T23:51:24.633-05:30');
});

test('Subtract Time 4', async () => {

    const subtractTimeFunction = await dfr.find(Namespaces.DATE, 'SubtractTime');

    if (!subtractTimeFunction) {
        throw new Error("Function not found");
    }
    
    fep.setArguments(new Map<string, any>([ 
        [AbstractDateFunction.PARAMETER_DATE_NAME, '2024-09-13T23:52:34.633-05:30'],
        [AbstractDateFunction.PARAMETER_INT_NAME, 5],
        [AbstractDateFunction.PARAMETER_UNIT_NAME, 'YEAR']
    ]));

    expect((await subtractTimeFunction.execute(fep)).allResults()[0].getResult().get('result')).toBe('2019-09-13T23:52:34.633-05:30');
});

File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/date/SubtractTimeTest.java
package com.fincity.nocode.kirun.engine.function.system.date;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.namespaces.Namespaces;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

public class SubtractTimeTest {
    
	DateFunctionRepository dfr = new DateFunctionRepository();

	ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
	        new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository());


    @Test
    void testInvalidDate(){


        fep.setArguments(Map.of("isoDate", new JsonPrimitive("2029-05-95T06:04:18.073Z") ,
        "intValue", new JsonPrimitive(10), "unit", new JsonPrimitive("MILLISECOND")));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SubtractTime")
                .flatMap(e -> e.execute(fep)))
                .expectError().verify();


        fep.setArguments(Map.of("isoDate", new JsonPrimitive(false),
        "intValue", new JsonPrimitive(10), "unit", new JsonPrimitive("MILLISECOND")));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SubtractTime")
                .flatMap(e -> e.execute(fep)))
                .expectError().verify();
    }

    @Test   
    void testSubtractTime1(){

		fep.setArguments(Map.of("isoDate", new JsonPrimitive("2024-09-13T23:52:34.633-05:30"),
        "intValue", new JsonPrimitive(10), "unit", new JsonPrimitive("MINUTE")));

	
        StepVerifier.create(dfr.find(Namespaces.DATE, "SubtractTime")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(res -> res.next()
                .getResult()
                .get("result")
                .getAsString().equals("2024-09-13T23:42:34.633-05:30"))
                .verifyComplete();

    }

    @Test
    void testSubtractTime2(){

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("2024-09-13T23:52:34.633-05:30"),
        "intValue", new JsonPrimitive(13), "unit", new JsonPrimitive("MONTH")));

	
        StepVerifier.create(dfr.find(Namespaces.DATE, "SubtractTime")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(res -> res.next()
                .getResult()
                .get("result")
                .getAsString().equals("2023-08-13T23:52:34.633-05:30"))
                .verifyComplete();

    }

    @Test
    void testSubtractTime3(){

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("2024-09-13T23:52:34.633-05:30"),
        "intValue", new JsonPrimitive(70), "unit", new JsonPrimitive("SECOND")));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SubtractTime")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(res -> res.next()
                .getResult()
                .get("result")
                .getAsString().equals("2024-09-13T23:51:24.633-05:30"))
                .verifyComplete();
        
    }

    @Test
    void testSubtractTime4(){


        fep.setArguments(Map.of("isoDate", new JsonPrimitive("2024-09-13T23:52:34.633-05:30"),
        "intValue", new JsonPrimitive(5), "unit", new JsonPrimitive("YEAR")));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SubtractTime")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(res -> res.next()
                .getResult()
                .get("result")
                .getAsString().equals("2019-09-13T23:52:34.633-05:30"))
                .verifyComplete();

    }

}
